src/main/java/org/broadinstitute/hellbender/tools/spark/sv/ContigAligner.java

```diff
@@ -0,0 +1,272 @@
+     * Inserted sequence:
+     *  GA
+     *
+     * Homology represents ambiguity about the exact location of the breakpoint. For example, in this case one alignment
+     * region ends with "AC" and the next begins with AC, so we don't know if the AC truly belongs with the first or
+     * second alignment region.
+     *
+     * Contig:
+     * ACTGACTGACTGACTGACTGACTGACTGACTGACTGACTGACTG
+     * Alignment regions:
+     * |-----1:100-200-------|
+     *                    |-----2:100-200----------|
+     * Homology:
+     *  AC
+     */
+    static class AssembledBreakpoint {
+        String contigId;
+        AlignmentRegion region1;
+        AlignmentRegion region2;
+        String insertedSequence;
+        String homology;
+
+        @Override
+        public String toString() {
+            final StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append(contigId);
+            stringBuilder.append("\t");
+            stringBuilder.append(region1.toString());
+            stringBuilder.append("\t");
+            stringBuilder.append(region2.toString());
+            stringBuilder.append("\t");
+            stringBuilder.append(insertedSequence);
+            stringBuilder.append("\t");
+            stringBuilder.append(homology);
+            return stringBuilder.toString();
+        }
+    }
+
+    static class AlignmentRegion {
+
+        final Cigar cigar;
+        final boolean forwardStrand;
+        final SimpleInterval interval;
+        final int mqual;
+        final int startInContig;
+        final int endInContig;
+        final int contigLength;
+
+
+        public AlignmentRegion(final AlnRgn alnRgn) {
+            this.forwardStrand = alnRgn.getStrand() == '+';
+            this.cigar = TextCigarCodec.decode(alnRgn.getCigar());
+            this.interval = new SimpleInterval(alnRgn.getChrom(), (int) alnRgn.getPos() + 1, (int) (alnRgn.getPos() + 1 + cigar.getReferenceLength()));
+            this.mqual = alnRgn.getMQual();
+            this.contigLength = cigar.getReadLength();
+            this.startInContig = startOfAlignmentInContig();
+            this.endInContig = endOfAlignmentInContig();
+        }
+
+        private int startOfAlignmentInContig() {
+            return getClippedBases(forwardStrand) + 1;
+        }
+
+        private int endOfAlignmentInContig() {
+            return contigLength - getClippedBases(!forwardStrand);
+        }
+
+        private int getClippedBases(final boolean forwardStrand) {
+            int posInContig = 0;
+            if (forwardStrand) {
+                int i = 0;
+                while (cigar.getCigarElement(i).getOperator() == CigarOperator.H || cigar.getCigarElement(i).getOperator() == CigarOperator.S) {
+                    posInContig += cigar.getCigarElement(i).getLength();
+                    i = i + 1;
+                }
+            } else {
+                int i = cigar.getCigarElements().size() - 1;
```

Comment:
You could consider using a ListIterator, rather than index arithmetic.  (It lets you step backward.)